materialize-boilerplate/boilerplate.go

```diff
@@ -70,6 +51,46 @@ type cmdCommon struct {
 	srv pm.DriverServer
 	r   *bufio.Reader
 	w   protoio.Writer
+
+	logging *logConfig
+}
+
+// logConfig configures handling of application log events.
+type logConfig struct {
+	Level  string `long:"log.level" env:"LOG_LEVEL" default:"info" choice:"info" choice:"INFO" choice:"debug" choice:"DEBUG" choice:"warn" choice:"WARN" description:"Logging level"`
+	Format string `long:"log.format" env:"LOG_FORMAT" default:"text" choice:"json" choice:"text" choice:"color" description:"Logging output format"`
+}
+
+func (c *cmdCommon) Execute(args []string) error {
+	if c.logging.Format == "json" {
+		logrus.SetFormatter(&logrus.JSONFormatter{})
+	} else if c.logging.Format == "text" {
+		logrus.SetFormatter(&logrus.TextFormatter{})
+	} else if c.logging.Format == "color" {
+		logrus.SetFormatter(&logrus.TextFormatter{ForceColors: true})
+	}
+
+	if lvl, err := logrus.ParseLevel(c.logging.Level); err != nil {
+		logrus.WithField("err", err).Fatal("unrecognized log level")
+	} else {
+		logrus.SetLevel(lvl)
+	}
+
+	const addr = ":2222"
+	logrus.WithField("addr", addr).Debug("waiting for gRPC connection")
+
+	var conn net.Conn
+	if server, err := net.Listen("tcp", addr); err != nil {
```

Comment:
Is it missing `defer server.Close()`? Same for client socket.

Reply:
It's not really necessary, the listening socket will remain open until the process exits, and it's the OS's responsibility to clean things up at that point.

Even so I'd normally prefer to include the cleanup logic, but in this particular case it'd be a bit of a pain to refactor things so that would work properly (because the command-line parsing library dispatches control to a bunch of distinct `Execute()` methods for each subcommand), so I'm happy just leaving it open.

Reply:
Ya, noticed flow is a bit complicated. Should be fine for short lived single instance runs but I'm not sure if we run these in parallel :)  Perhaps just add a comment to make it explicit.